Add an example gallery to the Universal Gateway section (#1365)
ngrok/ngrok-docs: docs/universal-gateway/examples/front-door-pattern.mdx

---
title: "Implement the 'Front Door' Pattern"
description: By using Cloud Endpoint and internal Agent Endpoints together, you loosely double your gateway and services for extra flexibility.
sidebar_label: "The Front Door Pattern"
---

import ReserveDomain from "./snippets/_reserve-domain.mdx";
import CloudEndpoint from "./snippets/_cloud-endpoint.mdx";
import Back from "./snippets/_back-to-examples.mdx";

The "front door" pattern involves using a Cloud Endpoint, the forward-internal Traffic Policy action, and an internal Agent Endpoint to route traffic from the public internet to your service(s).

With this pattern, you can:

- Host any number of services under a single hostname (e.g. `your-company.com`).

> - Decouple gateway configuration from per-service traffic management policies.

Reviewer: It might be useful in this sentence to describe what the benefit of this would be. As it stands, it feels really technical and hard to understand what the upshot of this is

> - Prevent any driect exposure of your services or their host systems to the public internet.

Reviewer: ```suggestion
- Prevent your services and their host systems from being exposed to the public internet.
```
Rewording this to put the subject at the start of the sentence (a little easier to parse IMO) and making it a little more succinct

- Route traffic and add authentication to microservices with a single API gateway configuration.
- Authenticate developer environments or CI preview builds behind a single endpoint.

## 1. Create an endpoint for your service

Start an internal [Agent Endpoint](/docs/universal-gateway/agent-endpoints/), replacing `$PORT` based on where your service listens.

> You can also use one of our [SDKs](/docs/agent/sdks) or the [Kubernetes Operator](/docs/k8s).

Reviewer: ```suggestion
You can also use one of our [SDKs](/docs/agent-sdks) or the [Kubernetes Operator](/docs/k8s).
```
Broken link

```bash
ngrok http $PORT --url https://service.internal
```

## 2. Reserve a domain

<ReserveDomain />

## 3. Create a Cloud Endpoint

<CloudEndpoint />

## 4. Add routing to your service with Traffic Policy

> Copy and paste the policy below into your new Cloud Endpoint in the ngrok dashboard.

Reviewer: ```suggestion
While still viewing your new cloud endpoint in the dashboard, copy and paste the policy below into the Traffic Policy editor.
```

```yaml
on_http_request:
  - actions:
      - type: forward-internal
        config:
          url: http://service.internal
```

### What's happening here?

> On every HTTP request, the Traffic Policy engine forwards the traffic to the

Reviewer: ```suggestion
On every HTTP request, the traffic policy engine forwards the traffic to the
```
I've likely missed it in a few places elsewhere in the guide, but our convention is to capitalized company products only [the first time it appears under](https://www.notion.so/ngrok/Style-Guide-1d6d19ffb65480a6ab31fca7a303030f?source=copy_link#1d6d19ffb65480a9a553cb370a95d519) an H2 (this is under an H3)

internal agent endpoint you created at `https://service.internal`.

> ## 5. Add a second service and routing (optional)

Reviewer: Let's move this and 5a out of the steps and turn step six into step five. Since this is optional, we shouldn't include it as part of the steps because it makes the experience fragmented

If you have another service to host under the front door pattern, start another agent.

```bash
ngrok http $PORT --url https://service-two.internal
```

## 5a. Update your routing based on paths

Copy and paste the policy below into the ngrok dashboard, replacing `/one` and `/two` with the paths you'd like to use for routing traffic to each service.

```yaml
on_http_request:
  - expressions:
      - "req.url.includes('/one')"
    actions:
      - type: forward-internal
        config:
          url: http://service.internal
  - expressions:
      - "req.url.includes('/two')"
    actions:
      - type: forward-internal
        config:
          url: http://service-two.internal
```

## 6. Try out your endpoint!

> - In your browser: https://your-service.ngrok.app
> - In your terminal: `curl https://your-service.ngrok.app`

Reviewer: ```suggestion
Visit your reserved domain either in the browser or in the terminal using a tool like `curl`. You should see the app or service at the port connected to your internal endpoint. 
```
Feel free to modify this suggestion. The spirit of it is that we don't know what their URL will be, and people skimming this might just copy this placeholder URL you've written, paste it in the browser, see nothing working, and get a bit confused for a second. I think it would be better to just explicitly say that they should use their reserved domain

## What's next?

> - Read more about [Traffic Policy](/docs/traffic-policy) and its [core concepts](/docs/traffic-policy/concepts).
> - Add authentication to your endpoint with the [`oauth` action](/docs/traffic-policy/actions/oauth) action or [`basic-auth` action](/docs/traffic-policy/actions/basic-auth).

Reviewer: ```suggestion
- [Traffic Policy](/docs/traffic-policy) enables you to do many more things worth your endpoints. For example, try adding authentication to your endpoint with the [`oauth` action](/docs/traffic-policy/actions/oauth) or [`basic-auth` action](/docs/traffic-policy/actions/basic-auth) actions.
```
I think we should condense these two into one point because just telling people to learn more about Traffic Policy doesn't give them a reason why, and so they'll be hesitant to click that link. But the second point was basically explaining why you should use traffic policy, so it's really one point.

> - View your traffic in [Traffic Inspector](https://dashboard.ngrok.com/traffic-inspector).

Reviewer: What's the benefit of this?

---
title: "Implement the 'Front Door' Pattern"
description: By using Cloud Endpoint and internal Agent Endpoints together, you loosely double your gateway and services for extra flexibility.
sidebar_label: "The Front Door Pattern"
---

import ReserveDomain from "./snippets/_reserve-domain.mdx";
import CloudEndpoint from "./snippets/_cloud-endpoint.mdx";
import TryOut from "./snippets/_try-out.mdx";
import Back from "./snippets/_back-to-examples.mdx";

> The "front door" pattern involves using a Cloud Endpoint, the forward-internal Traffic Policy action, and an internal Agent Endpoint to route traffic from the public internet to your service(s).

Reviewer 2: We should explain the actual pattern itself and why it is useful before introducing how it's done on ngrok.

- Apply certain policies like authentication at your gateway, then layer in other policies for specific services.
- Prevent your services or their host systems from being exposed to the public internet.
- Route traffic and add authentication to microservices with a single API gateway configuration.
- Authenticate developer environments or CI preview builds behind a single endpoint.

## 1. Create an endpoint for your service

Start an internal [Agent Endpoint](/docs/universal-gateway/agent-endpoints/), replacing `$PORT` based on where your service listens.
You can also use one of our [SDKs](/docs/agent-sdks) or the [Kubernetes Operator](/docs/k8s).

```bash
ngrok http $PORT --url https://service.internal
```

## 2. Reserve a domain

<ReserveDomain />

## 3. Create a Cloud Endpoint

<CloudEndpoint />

## 4. Add routing to your service with Traffic Policy

While still viewing your new cloud endpoint in the dashboard, copy and paste the policy below into the Traffic Policy editor.

```yaml
on_http_request:
  - actions:
      - type: forward-internal
        config:
          url: http://service.internal
```

> ### What's happening here?

Reviewer: This is a minor point, but I think this heading is probably unnecessary. We can just put the explanation after the code snippet without the heading. This applies generally everywhere we have the "What's happening here" heading
